File: digital_game.lua
--[[数字游戏--]]
local DigitalGame ={
    rows=5,
    cols=5,
    weights={0.35, 0.65, 0.85, 0.95, 1},
    numMap={}
 }
 function DigitalGame:randMap()
     --按概率生成数字
     for i=1,self.rows do
         self.numMap[i] = {}
         for j=1,self.cols do
             local rs = math.random()
             for k=1,#self.weights do
                 if rs<self.weights[k] then
                     self.numMap[i][j] = k
                     break
                 end
             end
         end
     end
     --能消除的元素 重新生成
     while  self:clearSameElement().bClear do
         self:randFill()
     end
     return self.numMap
 end
 --操作3个或者以上的相邻且相同的元素
 function  DigitalGame:clearSameElement()
     local rs = {
             bClear= false,
             clears= {}
         }
         --标记清除元素
         local function markCell(map, x, y, status)
             if x < 1 or x > self.rows  or y < 1  or y > self.cols  then
                 --越界的返回
                 return
             end
             if map[x][y] < 0 then
                 --已经访问过的返回
                 return
             end
             for i=1,#status.nodes do
                 local v = status.nodes[i]
                 if v.x == x and v.y==y then
                     --已经访问过的返回
                     return
                 end
             end
             if map[x][y] == status.value then
                 status.count = status.count+1
                 if status.count >= 3 then
                     if not status.bClear then
                         status.bClear = true
                         table.insert(rs.clears,{})
                         for i=1,#status.nodes do
                             local v = status.nodes[i]
                             map[v.x][v.y] = -map[v.x][v.y]
                             table.insert(rs.clears[#rs.clears],{v.x, v.y,status.value})
                         end
                     end
                     map[x][y] = -map[x][y]
                     table.insert(rs.clears[#rs.clears],{x, y,status.value})
                 else
                     table.insert(status.nodes,{x=x,y=y})
                 end
                 markCell(map, x - 1, y, status)
                 markCell(map, x + 1, y, status)
                 markCell(map, x, y - 1, status)
                 markCell(map, x, y + 1, status)
             end
         end
         for i=1,self.rows do
             for j=1,self.cols do
                 local status = {
                     value= self.numMap[i][j],
                     count= 0,
                     nodes= {},
                     bClear= false
                 }
                 markCell(self.numMap, i, j, status)
                 if status.bClear then
                     rs.bClear = true
                 end
             end
         end
         rs.map = self.numMap
         return rs
 end
 --填充函数---简单的随机填充
 function DigitalGame:randFill()
     for i=1,self.rows do
         for j=1,self.cols do
             if self.numMap[i][j] < 0 then
                 local rs = math.random()
                 for k=1,#self.weights do
                     if rs<self.weights[k] then
                         self.numMap[i][j] = k
                         break
                     end
                 end
             end
         end
     end
     return {
         map = self.numMap,
         actions = {}
     }
 end
 --填充函数---掉落填充，适用于消消乐
 function DigitalGame:dropFill()
         local actions = {}
         for j=1,self.cols do
             local exIdx = 0
             for i=self.rows,1,-1 do
                 if self.numMap[i][j] < 0 then
                     --往上寻找非0元素替换一下，相当于掉落
                     local bFind = false
                     for k=i-1,1,-1 do
                         if self.numMap[k][j] > 0 then
                             self.numMap[i][j], self.numMap[k][j] = self.numMap[k][j], self.numMap[i][j]
                             table.insert(actions,{k, j, i, j})
                             bFind = true
                             break
                         end
                     end
                     if not bFind then
                         --生成新的元素并掉落
                         local rs = math.random()
                         for k=1,#self.weights do
                             if rs<self.weights[k] then
                                 self.numMap[i][j] = k
                                 break
                             end
                         end
                         --超出边界，表示新生成元素从边界外掉落
                         table.insert(actions,{exIdx, j, i, j, self.numMap[i][j]})
                         exIdx = exIdx-1
                     end
                 end
             end
         end
         return {
             map= self.numMap,
             actions= actions
         }
 end
 --填充函数---掉落填充，合成点数字加1
 function DigitalGame:mergeFill(refPoint, clears)
         local pos = {}
         local filters = {}
         for j=1,self.cols do
             for i=self.rows,1,-1 do
                 if self.numMap[i][j] < 0 then
                     local bFind =false
                     for k=1,#filters do
                        local v=filters[k]
                        if v[1] == i and v[2] == j then
                             bFind=true
                             break
                        end
                     end
                     if not bFind then
                         for m=1,#clears do
                             local bFind =false
                             for n=1,#clears[m] do
                                 local v= clears[m][n]
                                 if v[1] == i and v[2] == j then
                                     bFind=true
                                 end
                             end
                             if bFind then
                                 for n=1,#clears[m] do
                                    table.insert(filters,clears[m][n])
                                 end
                             end
                         end
                         local cur = {i, j}
                         if refPoint then
                             cur = {refPoint.x, refPoint.y}
                         end
                         local oldVal = self.numMap[cur[1]][cur[2]]
                         local newVal = 1 - oldVal
                         table.insert(pos,{cur[1],cur[2],-oldVal,newVal})
                         self.numMap[cur[1]][cur[2]] = newVal
                     end
                 end
             end
         end
         local rs = self:dropFill()
         rs.mergePoints = pos
         return rs
 end
 function DigitalGame:check(refPoint)
     local rs = self:clearSameElement()
     local merges = self:mergeFill(refPoint,rs.clears)
     for k,v in pairs(merges) do
         rs[k]=v
     end
     return rs
 end

File: digital_game.js
//数字游戏。
////////////////////////////////////////////////////////////////////////////////////
//点击一次数字方块数字+1，且点击一次消耗1能量值				
//当相邻相同数字的方块数量大于3个（包含3个）时方块自动消除，并新生成一个消除方块数字+1的方块。								
//新生成方块显示位置：
//①消除的方块其中一个方块为玩家操作的则新生成的方块位置为操作方块位置，有多个操作方块则在最后操作方块的位置。
//②消除方块无有操作的方块新生成方块的位置最左最下的空位。（优先最左，左的位置一样选择最下）。
//③新生成的方块如果下的位置为空位，则向下掉落。
////////////////////////////////////////////////////////////////////////////////////
//生成新的数字游戏 
//@param {Number} rows 行数 
//@param {Number} cols 列数 
//@param {Array} weights 数字权重
//    例如:[0.35, 0.65, 0.85, 0.95, 1]表示1-5每个数字的生成概率为0.35,0.35,0.2,0.1,0.05
function DigitalGame(rows, cols, weights) {
    this.rows = rows
    this.cols = cols
    this.weights = weights
    this.numMap = []
    this.randMap()
}
DigitalGame.prototype = {
    randMap: function () {
        //按概率生成数字
        for (var i = 0; i < this.rows; i++) {
            this.numMap[i] = []
            for (var j = 0; j < this.cols; j++) {
                var rs = Math.random()
                this.numMap[i][j] = this.weights.findIndex(v => rs < v) + 1
            }
        }
        //能消除的元素 重新生成
        while (this.clearSameElement().bClear) {
            this.randFill()
        }
        return this.numMap
    },
    //操作3个或者以上的相邻且相同的元素
    clearSameElement: function () {
        var rs = {
            bClear: false,
            clears: []
        }
        //标记清除元素
        var markCell = (map, x, y, status) => {
            if (x < 0 || x > this.rows - 1 || y < 0 || y > this.cols - 1) {
                //越界的返回
                return
            }
            if (map[x][y] < 0 || status.nodes.find(v => v.x == x && v.y == y)) {
                //已经访问过的返回
                return
            }
            if (map[x][y] == status.value) {
                status.count += 1
                if (status.count >= 3) {
                    if (!status.bClear) {
                        status.bClear = true
                        rs.clears.push([])
                        status.nodes.forEach(v => {
                            map[v.x][v.y] *= -1
                            rs.clears[rs.clears.length - 1].push([v.x, v.y,status.value])
                        });
                    }
                    map[x][y] *= -1
                    rs.clears[rs.clears.length - 1].push([x, y,status.value])
                } else {
                    status.nodes.push({
                        x,
                        y
                    })
                }
                markCell(map, x - 1, y, status)
                markCell(map, x + 1, y, status)
                markCell(map, x, y - 1, status)
                markCell(map, x, y + 1, status)
            }
        }
        for (var i = 0; i < this.rows; i++) {
            for (var j = 0; j < this.cols; j++) {
                var status = {
                    value: this.numMap[i][j],
                    count: 0,
                    nodes: [],
                    bClear: false
                }
                markCell(this.numMap, i, j, status)
                if (status.bClear) {
                    rs.bClear = true
                }
            }
        }
        rs.map = this.numMap
        return rs
    },
    //填充函数---简单的随机填充
    randFill: function () {
        for (var i = 0; i < this.rows; i++) {
            for (var j = 0; j < this.cols; j++) {
                if (this.numMap[i][j] < 0) {
                    var rs = Math.random()
                    this.numMap[i][j] = this.weights.findIndex(v => rs < v) + 1
                }
            }
        }
        return {
            map: this.numMap,
            actions: []
        }
    },
    //填充函数---掉落填充，适用于消消乐
    dropFill: function () {
        var actions = []
        for (var j = 0; j < this.cols; j++) {
            var exIdx = -1
            for (var i = this.rows - 1; i >= 0; i--) {
                if (this.numMap[i][j] < 0) {
                    //往上寻找非0元素替换一下，相当于掉落
                    var bFind = false
                    for (var k = i - 1; k >= 0; k--) {
                        if (this.numMap[k][j] > 0) {
                            [this.numMap[i][j], this.numMap[k][j]] = [this.numMap[k][j], this.numMap[i][j]]
                            actions.push([k, j, i, j])
                            bFind = true
                            break
                        }
                    }
                    if (!bFind) {
                        //生成新的元素并掉落
                        var rs = Math.random()
                        this.numMap[i][j] = this.weights.findIndex(v => rs < v) + 1
                        //超出边界，表示新生成元素从边界外掉落
                        actions.push([exIdx, j, i, j, this.numMap[i][j]])
                        exIdx -= 1
                    }
                }
            }
        }
        return {
            map: this.numMap,
            actions: actions
        }
    },
    //填充函数---掉落填充，合成点数字加1
    mergeFill: function (refPoint, clears) {
        var pos = []
        var filters = []
        for (var j = 0; j < this.cols; j++) {
            for (var i = this.rows - 1; i > 0; i--) {
                if (this.numMap[i][j] < 0 && !filters.find(v => v[0] == i && v[1] == j)) {
                    for (var k = 0; k < clears.length; k++) {
                        if (clears[k].find(v => v[0] == i && v[1] == j)) {
                            filters.push(...clears[k])
                        }
                    }
                    var cur = [i, j]
                    if (refPoint) {
                        cur = [refPoint.x, refPoint.y]
                    }
                    var oldVal = this.numMap[cur[0]][cur[1]]
                    var newVal = 1 - oldVal
                    pos.push([...cur, -oldVal, newVal])
                    this.numMap[cur[0]][cur[1]] = newVal
                }
            }
        }
        var rs = this.dropFill()
        rs.mergePoints = pos
        return rs
    },
    check(refPoint) {
        var rs = this.clearSameElement()
        Object.assign(rs, this.mergeFill(refPoint, rs.clears))
        return rs
    }
}